torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -859,3 +859,58 @@ def _clear_grads_if_needed(
     for handle in handles:
         if handle._use_orig_params:
             handle._clear_grads_if_needed()
+
+
+@no_type_check
+def _get_buffers_and_dtypes_for_computation(
+    state: _State,
+    root_module: nn.Module,
+) -> Tuple[List[torch.Tensor], List[Optional[torch.dtype]]]:
+    """
+    Returns all buffers in the module tree rooted at ``root_module`` and a
+    corresponding list of the buffer dtypes for computation. Each buffer dtype
+    is either ``None`` if buffer mixed precision is not enabled or the buffer
+    low precision dtype otherwise.
+    """
+    buffers: List[torch.Tensor] = []
+    buffer_dtypes: List[Optional[torch.dtype]] = []
+    if _is_composable(state):
+        buffers = [
+            buffer for module in root_module.modules() for buffer in module.buffers()
+        ]
+        buffer_dtypes = [
+            state.mixed_precision.buffer_dtype for _ in range(len(buffers))
+        ]
+    else:
+        for fsdp_module in state.fsdp_modules(root_module):
```

Comment:
question: I guess this will be merged with composable path once non-recursive wrapping is landed next year?

Reply:
For the `_is_composable(state)` path above, we assume that `state.mixed_precision` defines the mixed precision setting for the entire module tree for which composable FSDP was applied.

The reason I have this `else` branch is that for existing FSDP, it is technically possible for different buffer mixed precision settings to be applied throughout the module tree.

I agree that we would be able to merge these paths if we only have top-level `FullyShardedDataParallel`. The precise condition is that if we can assume the same mixed precision setting for the entire module tree, then we can merge the paths.